test/unit/buffer_spec.lua

```diff
@@ -211,93 +212,177 @@ describe('buffer functions', function()
     end
 
-    it('should copy plain text', function()
-      local width = build_stl_str_hl("this is a test")
-
-      eq(14, width)
-      eq("this is a test", helpers.ffi.string(output_buffer, width))
-
-    end)
-
-    it('should print no file name', function()
-      local width = build_stl_str_hl("%f")
-
-      eq(9, width)
-      eq("[No Name]", helpers.ffi.string(output_buffer, width))
-
-    end)
-
-    it('should print the relative file name', function()
-      buffer.setfname(globals.curbuf, to_cstr("Makefile"), NULL, 1)
-      local width = build_stl_str_hl("%f")
-
-      eq(8, width)
-      eq("Makefile", helpers.ffi.string(output_buffer, width))
-
-    end)
-
-    it('should print the full file name', function()
-      buffer.setfname(globals.curbuf, to_cstr("Makefile"), NULL, 1)
-
-      local width = build_stl_str_hl("%F")
-
-      assert.is_true(8 < width)
-      neq(NULL, string.find(helpers.ffi.string(output_buffer, width), "Makefile"))
-
-    end)
-
-    it('should print the tail file name', function()
-      buffer.setfname(globals.curbuf, to_cstr("src/nvim/buffer.c"), NULL, 1)
-
-      local width = build_stl_str_hl("%t")
-
-      eq(8, width)
-      eq("buffer.c", helpers.ffi.string(output_buffer, width))
-
-    end)
-
-    it('should print the buffer number', function()
-      buffer.setfname(globals.curbuf, to_cstr("src/nvim/buffer.c"), NULL, 1)
-
-      local width = build_stl_str_hl("%n")
-
-      eq(1, width)
-      eq("1", helpers.ffi.string(output_buffer, width))
-    end)
-
-    it('should print the current line number in the buffer', function()
-      buffer.setfname(globals.curbuf, to_cstr("test/unit/buffer_spec.lua"), NULL, 1)
-
-      local width = build_stl_str_hl("%l")
-
-      eq(1, width)
-      eq("0", helpers.ffi.string(output_buffer, width))
-
-    end)
-
-    it('should print the number of lines in the buffer', function()
-      buffer.setfname(globals.curbuf, to_cstr("test/unit/buffer_spec.lua"), NULL, 1)
-
-      local width = build_stl_str_hl("%L")
+    -- Use this function to simplify testing the comparison between
+    --  the format string and the resulting statusline.
+    --
+    -- @param description The description of what the test should be doing
+    -- @param statusline_cell_count The number of cells available in the statusline
+    -- @param expected_cell_count The expected number of cells build_stl_str_hl will return
+    -- @param expected_byte_length The expected byte length of the string
+    -- @param input_stl The format string for the statusline
+    -- @param expected_stl The expected result string for the statusline
+    local function statusline_test (description,
+                                    statusline_cell_count,
+                                    expected_cell_count,
+                                    expected_byte_length,
+                                    input_stl,
+                                    expected_stl,
+                                    arg)
+
```

Comment:
I do not like the idea of the function with seven arguments. While it is easy to understand what most of the arguments mean, meanings of the three numbers in the middle are hard to remember. I also think that expected_byte_length may default to expected_cell_count which may default to statusline_cell_count, so some tests will have one or two arguments less.